lib/runtime/StartupChunkDependenciesPlugin.js

```diff
@@ -80,10 +90,66 @@ class StartupChunkDependenciesPlugin {
 							chunk,
 							new StartupEntrypointRuntimeModule(this.asyncChunkLoading)
 						);
-					});
-			}
+					},
+				);
+				
+				const { renderStartup } =
+          		compiler.webpack.javascript.JavascriptModulesPlugin.getCompilationHooks(
+           		 compilation,
+          		);
+
+        		renderStartup.tap(
+          		'MfStartupChunkDependenciesPlugin',
+          		(startupSource, lastInlinedModule, renderContext) => {
+            	const { chunk, chunkGraph, runtimeTemplate } = renderContext;
+
+            	if (!isEnabledForChunk(chunk)) {
+              return startupSource;
+            	}
+
+            	const treeRuntimeRequirements =
+              chunkGraph.getTreeRuntimeRequirements(chunk);
+            	const chunkRuntimeRequirements =
+              chunkGraph.getChunkRuntimeRequirements(chunk);
+
+            	const federation =
+              chunkRuntimeRequirements.has(federationStartup) ||
+              treeRuntimeRequirements.has(federationStartup);
+
+            	if (!federation) {
+              return startupSource;
+            	}
+
+            	const entryModules = Array.from(
+              chunkGraph.getChunkEntryModulesWithChunkGroupIterable(chunk),
+           	 );
+
+           	 const entryGeneration = runtimeTemplate.outputOptions.module
+              // @ts-ignore
```

Comment:
ESM entry has its own startup. Instead of adding another render startup hook, etc., we should adjust the existing ESM startup and startup chunks plugin, etc.